### PR TITLE
Fix contentInsetAdjustmentBehavior set to automatic on ScrollView in the new architecture

### LINKED ISSUE #34165: ScrollView doesn't reset inset when contentInsetAdjustmentBehavior is automatic
### Description

This is similar to but **NOT** the same as the [bug I raised in the Working Group](https://github.com/reactwg/react-native-new-architecture/discussions/30). The difference here is that the ScrollView has scrollable content. 

I've started updating my navigation library to work with the new architecture and I've run into a problem on iOS. The `ScrollView` doesn't work when `contentInsetAdjustmentBehavior` is set to `automatic`. You can see what I mean in the first video below. The first time you navigate to the second screen then the ScrollView behaves correctly. But the second time you navigate you can see that the `ScrollView` isn't adjusting for the navigation bar. The second video is the same example but with the old architecture. You can see that the `ScrollView` behaves correctly.

All these bugs are because the new architecture is recycling the `ScrollView`. I turned off the recycling of the `ScrollView` and the bug went away. Would you be open to a PR that turns off view recycling for `ScrollViews` with adjustment behaviour set to automatic?
```obj-c
// In _enqueueComponentViewWithComponentHandle of RCTComponentViewRegistry.mm
BOOL recycle = ![componentViewDescriptor.view isKindOfClass:[RCTScrollViewComponentView class]]
    || ((RCTScrollViewComponentView *) componentViewDescriptor.view).scrollView.contentInsetAdjustmentBehavior == UIScrollViewContentInsetAdjustmentAutomatic;
if (recycle)
  recycledViews.push_back(componentViewDescriptor);
```
It should be possible to keep recycling on for the `RCTSCrollViewComponent` and just create the `UIScrollView` new each time. But I tried it and couldn't get it to work.

### Version

0.69.1

### Output of `npx react-native info`

System:
    OS: macOS 12.4
    CPU: (8) arm64 Apple M1 Pro
    Memory: 461.39 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 17.7.1 - /opt/homebrew/bin/node
    Yarn: 1.22.18 - /opt/homebrew/bin/yarn
    npm: 8.5.2 - /opt/homebrew/bin/npm
    Watchman: 2022.03.14.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    Android SDK:
      API Levels: 29, 30, 31, 32
      Build Tools: 29.0.2, 30.0.2, 31.0.0, 32.0.0, 32.1.0
      System Images: android-30 | Intel x86 Atom_64, android-30 | Google Play ARM 64 v8a, android-32 | Google APIs ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2021.1 AI-211.7628.21.2111.8193401
    Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.11 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.0.0 => 18.0.0 
    react-native: 0.69.1 => 0.69.1 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found

### Steps to reproduce

I've created a [repository with a reproduction of the bug](https://github.com/grahammendick/content-inset-adjustment-auto-scroll-bug). The README in the repo contains the steps needed to recreate the problem. There are two apps in the repo, one built with the old architecture and the other built with the new architecture. You can see that the bug only happens on the new architecture. The only dependencies in the apps are the 3 packages that make up my navigation library.

The ScrollView misbehaving on the new architecture

https://user-images.githubusercontent.com/1761227/178107788-8c64111a-3862-40e3-9f62-7e36c188c85d.mov

The ScrollView behaving correctly on the old architecture

https://user-images.githubusercontent.com/1761227/178107790-2d77ce3e-5142-4f3d-9716-e2b73eac479b.mov

### Snack, code example, screenshot, or link to a repository

https://github.com/grahammendick/content-inset-adjustment-auto-scroll-bug

### PR DESCRIPTION
**Pull request by grahammendick**
## Summary

Fixes #34165 and [Large title fails](https://github.com/reactwg/react-native-new-architecture/discussions/43) on the new React Native architecture.

There are problems with setting `contentInsetAdjustmentBehavior` to `automatic` on the `ScrollView` component in the new React Native architecture. The `automatic` setting matters to navigation libraries (like [my Navigation router](https://github.com/grahammendick/navigation)) because it stops the `ScrollView` from overlapping the `UINavigationBar`. The setting also powers important native features like large titles and search bars on iOS.

The `automatic` setting works fine on the old architecture. It doesn’t work on the new architecture because React Native is recycling views. In #34165 and [Large title fails](https://github.com/reactwg/react-native-new-architecture/discussions/43) there are videos comparing the setting in the old and the new architecture.

## Changelog

[iOS] [Fixed] - Fix `contentInsetAdjustmentBehavior` set to `automatic` on `ScrollView` in the new architecture

## Test Plan

I checked the fix in both the repros in #34165 and [Large title fails](https://github.com/reactwg/react-native-new-architecture/discussions/43). Here is a video of the fix running with large titles in the new architecture.

https://user-images.githubusercontent.com/1761227/179612188-162b896b-82c5-45de-bb5a-ba80f452fbee.mov

### CODE REVIEW

File: React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm [hunk 1]
```diff
@@ -130,6 +131,7 @@ - (instancetype)initWithFrame:(CGRect)frame
     [self.scrollViewDelegateSplitter addDelegate:self];
 
     _scrollEventThrottle = INFINITY;
+    _oldFrame = CGRectNull;
```
Comment (cipolleschi) [inline React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm:134]: I think is better if we use `CGRectZero` here

File: React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm [hunk 2]
```diff
@@ -401,6 +407,8 @@ - (void)prepareForRecycle
   _shouldUpdateContentInsetAdjustmentBehavior = YES;
   _state.reset();
   _isUserTriggeredScrolling = NO;
+  _oldFrame = self.frame;
+  [self setFrame:CGRectZero];
```
Comment (cipolleschi) [inline React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm:411]: Could you try to replace this part with this code:

```objc
  CGRect oldFrame = self.frame;
  self.frame = CGRectZero;
  self.frame = oldFrame;
```

and remove all the other changes. It should work and it save us from keeping a state variable which can get out of sync easily.

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi @grahammendick, thanks for the PR. 
I left some suggestions that could simplify the code. Thanks to @sammy-SC for them.

If you could try the last change, it could make the code much easier.

File: React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm (hunk 2 above)
Comment (grahammendick) [reply React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm]: Brilliant. I should've thought of that 👍

**Comment by sammy-SC** [comment]
@grahammendick thank you for fixing this!

**Comment by grahammendick** [comment]
@sammy-SC it's a crucial fix for my Navigation router so thanks for merging it so quickly 👍

**Pull request closed without merging**